components/feature/session/src/main/java/mozilla/components/feature/session/prompt/ChoiceAdapter.kt

```diff
@@ -0,0 +1,140 @@
+package mozilla.components.feature.session.prompt
+
+import android.support.v7.widget.RecyclerView
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import android.widget.CheckedTextView
+import android.widget.TextView
+import mozilla.components.concept.engine.prompt.Choice
+import mozilla.components.feature.session.R
+import mozilla.components.feature.session.prompt.ChoiceAdapter.ChoiceViewHolder
+import java.lang.IllegalArgumentException
+
+/**
+ * RecyclerView adapter for displaying choice items.
+ */
+internal class ChoiceAdapter(
+    private val fragment: ChoiceDialogFragment,
+    private val inflater: LayoutInflater
+) : RecyclerView.Adapter<ChoiceViewHolder>() {
+
+    companion object {
+        internal const val TYPE_MULTIPLE = 1
+        internal const val TYPE_SINGLE = 2
+        internal const val TYPE_GROUP = 3
+        internal const val TYPE_MENU = 4
+        internal const val TYPE_SEPARATOR = 5
+    }
+
+    private val choices = mutableListOf<Choice>()
+
+    init {
+        addItems(fragment.choices)
+    }
+
+    override fun getItemViewType(position: Int): Int {
+        val item = choices[position]
+        return when {
+            fragment.isSingleChoice -> TYPE_SINGLE
+            fragment.isMenuChoice -> if (item.isASeparator) TYPE_SEPARATOR else TYPE_MENU
+            item.isGroupType -> TYPE_GROUP
+            else -> TYPE_MULTIPLE
+        }
+    }
+
+    override fun onCreateViewHolder(parent: ViewGroup, type: Int): ChoiceViewHolder {
+        return ChoiceViewHolder(inflater.inflate(getLayoutId(type), parent, false))
+    }
+
+    override fun getItemCount(): Int = choices.size
+
+    @Suppress("ComplexMethod")
+    override fun onBindViewHolder(holder: ChoiceViewHolder, position: Int) {
```

Comment:
This is indeed a complex method. I wonder if it would be simpler if we would use multiple view holders types?

Reply:
Fixed. Yes, it really helped!